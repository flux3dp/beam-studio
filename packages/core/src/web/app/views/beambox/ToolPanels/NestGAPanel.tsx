import React from 'react';

import UnitInput from 'app/widgets/Unit-Input-v2';

interface Props {
  nestOptions: {
    generations: number,
    population: number,
  }
  updateNestOptions: (options: {
    generations?: number,
    population?: number,
  }) => void,
}

function NestGAPanel({ nestOptions, updateNestOptions }: Props): JSX.Element {
  const [generations, updateGenerations] = React.useState(nestOptions.generations);
  const [population, updatePopulation] = React.useState(nestOptions.population);

  const updateGen = (val: number) => {
    updateNestOptions({ generations: val });
    updateGenerations(val);
  };

  const updatePopu = (val: number) => {
    updateNestOptions({ population: val });
    updatePopulation(val);
  };

  return (
    <div className="tool-panel">
      <label className="controls accordion">
        <input type="checkbox" className="accordion-switcher" defaultChecked />
        <p className="caption">
          GA
          <span className="value">{`G${generations}, P${population}`}</span>
        </p>
        <label className="accordion-body">
          <div>
            <span className="text-center header">Generations</span>
            <div className="control">
              <UnitInput
                min={1}
                unit=""
                decimal={0}
                defaultValue={generations}
                getValue={updateGen}
              />
            </div>
            <span className="text-center header">Population</span>
            <div className="control">
              <UnitInput
                min={2}
                unit=""
                decimal={0}
                defaultValue={population}
                getValue={updatePopu}
              />
            </div>
          </div>
        </label>
      </label>
    </div>
  );
}

export default NestGAPanel;
