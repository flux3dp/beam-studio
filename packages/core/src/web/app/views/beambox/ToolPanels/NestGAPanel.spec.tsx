import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import NestGAPanel from './NestGAPanel';

jest.mock('app/widgets/Unit-Input-v2', () =>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ({ min, defaultValue, getValue }: any) => (
    <div>
      mock-unit-input min:{min}
      defaultValue:{defaultValue}
      <input className="unit-input" onChange={(e) => getValue(+e.target.value)} />
    </div>
  )
);

test('should render correctly', () => {
  const updateNestOptions = jest.fn();
  const { container } = render(
    <NestGAPanel
      nestOptions={{
        generations: 3,
        population: 10,
      }}
      updateNestOptions={updateNestOptions}
    />
  );
  expect(container).toMatchSnapshot();

  const inputs = container.querySelectorAll('input.unit-input');
  fireEvent.change(inputs[0], { target: { value: 2 } });
  expect(updateNestOptions).toHaveBeenCalledTimes(1);
  expect(updateNestOptions).toHaveBeenNthCalledWith(1, { generations: 2 });
  expect(container).toMatchSnapshot();

  fireEvent.change(inputs[1], { target: { value: 9 } });
  expect(updateNestOptions).toHaveBeenCalledTimes(2);
  expect(updateNestOptions).toHaveBeenNthCalledWith(2, { population: 9 });
  expect(container).toMatchSnapshot();
});
