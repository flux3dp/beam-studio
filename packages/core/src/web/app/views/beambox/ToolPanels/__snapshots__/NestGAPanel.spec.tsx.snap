// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    class="tool-panel"
  >
    <label
      class="controls accordion"
    >
      <input
        checked=""
        class="accordion-switcher"
        type="checkbox"
      />
      <p
        class="caption"
      >
        GA
        <span
          class="value"
        >
          G3, P10
        </span>
      </p>
      <label
        class="accordion-body"
      >
        <div>
          <span
            class="text-center header"
          >
            Generations
          </span>
          <div
            class="control"
          >
            <div>
              mock-unit-input min:
              1
              defaultValue:
              3
              <input
                class="unit-input"
              />
            </div>
          </div>
          <span
            class="text-center header"
          >
            Population
          </span>
          <div
            class="control"
          >
            <div>
              mock-unit-input min:
              2
              defaultValue:
              10
              <input
                class="unit-input"
              />
            </div>
          </div>
        </div>
      </label>
    </label>
  </div>
</div>
`;

exports[`should render correctly 2`] = `
<div>
  <div
    class="tool-panel"
  >
    <label
      class="controls accordion"
    >
      <input
        checked=""
        class="accordion-switcher"
        type="checkbox"
      />
      <p
        class="caption"
      >
        GA
        <span
          class="value"
        >
          G2, P10
        </span>
      </p>
      <label
        class="accordion-body"
      >
        <div>
          <span
            class="text-center header"
          >
            Generations
          </span>
          <div
            class="control"
          >
            <div>
              mock-unit-input min:
              1
              defaultValue:
              2
              <input
                class="unit-input"
              />
            </div>
          </div>
          <span
            class="text-center header"
          >
            Population
          </span>
          <div
            class="control"
          >
            <div>
              mock-unit-input min:
              2
              defaultValue:
              10
              <input
                class="unit-input"
              />
            </div>
          </div>
        </div>
      </label>
    </label>
  </div>
</div>
`;

exports[`should render correctly 3`] = `
<div>
  <div
    class="tool-panel"
  >
    <label
      class="controls accordion"
    >
      <input
        checked=""
        class="accordion-switcher"
        type="checkbox"
      />
      <p
        class="caption"
      >
        GA
        <span
          class="value"
        >
          G2, P9
        </span>
      </p>
      <label
        class="accordion-body"
      >
        <div>
          <span
            class="text-center header"
          >
            Generations
          </span>
          <div
            class="control"
          >
            <div>
              mock-unit-input min:
              1
              defaultValue:
              2
              <input
                class="unit-input"
              />
            </div>
          </div>
          <span
            class="text-center header"
          >
            Population
          </span>
          <div
            class="control"
          >
            <div>
              mock-unit-input min:
              2
              defaultValue:
              9
              <input
                class="unit-input"
              />
            </div>
          </div>
        </div>
      </label>
    </label>
  </div>
</div>
`;
